### ISSUE TITLE
Metrics Payload Array

### ISSUE THREAD

**Issue by mattevans** [snikch/aggro#5]
### Description

At the moment, if we make a query for more than 1 metric, the first metric (primary) would be used within tables, charts, etc. Any subsequent metrics would be exposed to the user via tooltips upon hovering on chart/table data.

Currently, metrics are returned as...

```
{
  buckets: {
    'Full-Time': {
      value: 'Full-Time',
      metrics: {
        'job.id:count': 4, <--- PLOTTED ON CHART
        'job.salary:sum': 380000 <--- USED IN CHART TOOLTIP
      }
    }, 
    'Part-Time': {
      value: 'Part-Time',
      metrics: {
        'job.id:count': 7,
        'job.salary:sum': 625000
      }
    }
  }
}
```

From Chris:

> I’m finding it a lil hard working with so many associative arrays in the response. JavaScript isn’t so good at getting the first item in an associative array without knowing the key to get it by :disappointed:. Any chance we could change the structure of the response to just be an array, with the key set as a `key` property?

Example:

```
{
  buckets: [
    {
      value: 'Full-Time',
      metrics: [
        {
          key: 'job.id:count',
          value: 4
        },
        {
          key: 'job.salary:sum',
          value: 380000
        }
      ]
    }, {
      value: 'Part-Time',
      metrics: [
        {
          key: 'job.id:count',
          value: 7
        },
        {
          key: 'job.salary:sum',
          value: 625000
        }
      ]
    }
  ]
}
```

What's your thoughts changing the payload to handle this for Chris? I would have thought accessing the first element of the metrics object would be alright as it stands, but I haven't touched any JS in a while. 

I understand these payloads were designed with FE use in-mind (easy iteration over result set for display, etc), so any feedback before changes are made would be great.

**Comment by snikch** [comment]
Hey, so yea - I think for the bucket values there's no need for them to be maps rather than arrays. They literally already have the value there in the `value` key so it's not a problem. I'm not so keen on the metrics however, because those are lookups and are inherently unordered. For each result, both of those values will exist, and he should be asking for them based on another value, perhaps the jsonified list of metrics that were required in the first place.

For example if the query definition can be represented as the following js

```js
var queryDef = {
  "Bucket": {
    "Field": {
      "Name": "job_type",
      "Type": "string"
    }
  },
  "Metrics": [{
    "Type": "count",
    "Field": "job.id"
  }, {
    "Type": "sum",
    "Field": "job.salary"
  }]
}
```

Then you know you'd be able to, for each data cell, be able to loop over `queryDef.Metrics` and generate the keys.

```js
var keys = []
queryDef.metrics.each(function (metric) {
  keys.push(metric.Field + ":" + metric.Type)
})
```

Which you could then loop over with each data result. So if the data results looked like this.

```js
var results = {
  buckets: [{
      value: 'Full-Time',
      metrics: {
        'job.id:count': 4,
        'job.salary:sum': 380000
      }
    }, {
      value: 'Part-Time',
      metrics: {
        'job.id:count': 7,
        'job.salary:sum': 625000
      }
    }
  ]
}
```

Then you'd easily be able to loop over the results and access the metrics

```js
results.each(function (result) {
  keys.each(function (key) {
    console.log(result.value, result.metrics[key])
  })
})
// Full-Time 4
// Full-Time 380000
// Part-time 7
// Part-time 625000
```

It's even easier in the case that you know that one value is the display value and another is a tooltip - since you're looking up each value specifically as if you have a level of "prior knowledge".

One thing that could perhaps be useful is including the query that was run as part of the result set, but really the client has a copy of it anyway since they sent it with the request.

So tl;dr yes to bucket values, no to metrics. Make sense?

**Comment by mattevans** [comment]
Sorry about the slow response here - many things on-the-go at once.
Chris has gone with your recommendations here.
